Cookbook/Proposals/MealyReactiveFeedback.md

```diff
@@ -0,0 +1,273 @@
+
+However, from practical point of view, Mealy machine has several advantages over Moore machine:
+
+- Requires less states
+    - This is a general conclusion of Mealy machine compared to Moore because we will add more graph-edges (transition patterns) than graph-nodes (states).
+- `Event`-recognizable `Feedback`
+    - For simple side-effects such as **logging**, **sending analytics events**, or somtimes **routing** or **image-loading** can be handled just from incoming events without precisely handling their states inside `ReactiveFeedback`
+    - This will remove verbose "idle -> routing -> idle" navigation state transition dance, and improves the feedback-based-navigation idea in [Babylonpartners/ios-playbook#44](https://github.com/Babylonpartners/ios-playbook/pull/44)
+    - We can discard `Event`s to turn the system back into current Moore machine at any time (**backward compatible**)
+
+### 2. Distinguish between "external event" and "feedback"
+
+Current `Feedback` is NOT actually handling "feedback" only, but also handling "external events" by injecting them into feedback closure, e.g.:
+
+```swift
+let incrementFeedback = Feedback<Int, Event> { _ in // NOTE: argument is not used
+    incrementSignal.map { _ in Event.increment }
+}
+```
+
+Notice that `Signal<State>` argument is not used at all.
+While this code works in general, mixing these 2 different semantics causes verbosity and sometimes confusion.
+
+The ideal approach will be to **separate "external events" as plain `Signal<Event>`, and pass it as a new argument next to `Feedback`s**.
+
+The idea of "external event" has been previously discussed in [Babylonpartners/ReactiveFeedback#32](https://github.com/Babylonpartners/ReactiveFeedback/pull/32), and it also makes sense from automata theory that "external events" (alphabets) and "feedbacks" (output function) are different.
+Also, Elm provides [`Subscription`](https://package.elm-lang.org/packages/elm/core/latest/Platform-Sub) which works just as same as observing "external events" and it is different from feedback type called `Cmd<Event>`.
+
+Furthermore, by separating "external event" nicely, we can send "initial external event" only after system is instantiated, so that the racing issue described in [ReactiveFeedback/pull#38](https://github.com/Babylonpartners/ReactiveFeedback/pull/38#issuecomment-468478325) will no longer occur.
+
+### 3. Remove duplicated `state` caching in both `SignalProducer` and `Property`
+
+`State` is stored in both `scan` and `Property` which is obviously inefficient.
+
+## Proposed solution
+
+### Changes
+
+```diff
+ public struct Feedback<State, Event> {
+-    let events: (Scheduler, Signal<State, NoError>) -> Signal<Event, NoError>
++    let events: (Scheduler, Signal<(Event, State), NoError>) -> Signal<Event, NoError>
+ }
+
+ extension Property {
```

Comment:
Not sure if it is valuable to have `Property<(State, Event)>`.

Reply:
`Property<State>` stays as is, so it won't be `Property<(State, Event)>`.